File: bizlogic/arxiv_bizlogic/validation/password_validator.py
"""Password validator using precomputed bad password hashes."""

import hashlib
import functools
import json
import os

# Load hashes from JSON file
_HASH_FILE = os.path.join(os.path.dirname(__file__), 'bad_passwords_hashes.json')
with open(_HASH_FILE, 'r') as f:
    _hash_data = json.load(f)
    BAD_PASSWORD_HASHES = set(_hash_data['hashes'])


def is_bad_password(password: str) -> bool:
    """
    Check if a password is in the bad password list.

    Args:
        password: The password to validate (case-insensitive)

    Returns:
        True if password is in the bad list, False otherwise
    """
    password_hash = hashlib.sha256(password.lower().encode()).hexdigest()[:8]
    return password_hash in BAD_PASSWORD_HASHES


def validate_password_strength(password: str) -> tuple[bool, str]:
    """
    Validate password against bad password list.

    Args:
        password: The password to validate

    Returns:
        Tuple of (is_valid, error_message)
    """
    if len(password) < 8:
        return False, "Too short"

    counts = [0] * 65536
    for c in password:
        if ord(c) < 65536:
            counts[ord(c)] += 1
    max_counts = functools.reduce(max, counts)
    if max_counts >= len(password)/2:
        return False, "Too many repeated characters"

    if is_bad_password(password):
        return False, "This password is too common and not allowed"

    return True, ""


# Example usage
if __name__ == "__main__":
    test_passwords = ["password123", "MyS3cur3P@ss!", "12345678", "unique_password_2024"]

    for pwd in test_passwords:
        is_bad = is_bad_password(pwd)
        print(f"{pwd:30} -> {'BAD' if is_bad else 'OK'}")

File: bizlogic/arxiv_bizlogic/validation/passwordValidator.ts
/**
 * Password validator using precomputed bad password hashes.
 *
 * Usage:
 *   import { isBadPassword, validatePasswordStrength } from './passwordValidator';
 *
 *   if (isBadPassword('password123')) {
 *     console.log('Password is too common');
 *   }
 */

import * as crypto from 'crypto';
import badPasswordData from './bad_passwords_hashes.json';

// Create a Set for O(1) lookup
const BAD_PASSWORD_HASHES = new Set(badPasswordData.hashes);

/**
 * Check if a password is in the bad password list.
 *
 * @param password - The password to validate (case-insensitive)
 * @returns true if password is in the bad list, false otherwise
 */
export function isBadPassword(password: string): boolean {
  const hash = crypto
    .createHash('sha256')
    .update(password.toLowerCase())
    .digest('hex')
    .substring(0, 8);

  return BAD_PASSWORD_HASHES.has(hash);
}

/**
 * Validate password against bad password list.
 *
 * @param password - The password to validate
 * @returns Object with isValid flag and optional error message
 */
export function validatePasswordStrength(password: string): {
  isValid: boolean;
  error?: string;
} {
  if (isBadPassword(password)) {
    return {
      isValid: false,
      error: 'This password is too common and not allowed'
    };
  }

  return { isValid: true };
}

// Browser-compatible version (no crypto module)
export function isBadPasswordBrowser(password: string): Promise<boolean> {
  return crypto.subtle
    .digest('SHA-256', new TextEncoder().encode(password.toLowerCase()))
    .then(hashBuffer => {
      const hashArray = Array.from(new Uint8Array(hashBuffer));
      const hashHex = hashArray
        .map(b => b.toString(16).padStart(2, '0'))
        .join('')
        .substring(0, 8);

      return BAD_PASSWORD_HASHES.has(hashHex);
    });
}

// Example usage
if (require.main === module) {
  const testPasswords = ['password123', 'MyS3cur3P@ss!', '12345678', 'unique_password_2024'];

  console.log('Password Validation Results:');
  testPasswords.forEach(pwd => {
    const isBad = isBadPassword(pwd);
    console.log(`${pwd.padEnd(30)} -> ${isBad ? 'BAD' : 'OK'}`);
  });
}
